### PR TITLE
Disable changing distribution keys implicitly when creating unique index

### PR DESCRIPTION
**Pull request by gfphoenix78**
Discussed in https://groups.google.com/a/greenplum.org/g/gpdb-dev/c/BGoCv_KFuss

In previous GPDB version, the distribution keys may be changed implicitly
when creating a unique index on a hash-distributed empty table.
```SQL
create table foo(a int, b int) distributed by(a);
create unique index on foo(b);
-- now, foo is hash distributed by b, not a
```
It might be useful(maybe) to avoid changing the distribution keys. However,
on the other side, it's crazy if the user doesn't notice the NOTICE message
like, "NOTICE:  updating distribution policy to match new UNIQUE index".

What's worse, this behavior could bring data inconsistency. See,
```SQL
create table foo(a int, b int) distributed by(a);
insert into foo select i,i from generate_series(1,5)i;

create table foopart (i int4, j int4) distributed by (i) partition by
        range (i) (start (1) end (3) every (1));
create unique index on foopart_1_prt_1 (j);
insert into foopart values(1,2),(2,1);
```
The data inconsistency is
```
gpadmin=# select gp_segment_id, * from foopart_1_prt_1;
 gp_segment_id | i | j
 ---------------+---+---
             1 | 1 | 2
 (1 row)

gpadmin=# select * from foo f, foopart_1_prt_1 p where f.a = p.j;
 a | b | i | j
 ---+---+---+---
 (0 rows)
```

Implicitly changing the distribution keys is not very useful, but harmful.
This PR disables changing the distribution keys when creating unique index.
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by zhangh43** [comment]
@gfphoenix78 

I'm wondering whether we should disable this feature totally. As Heikki mentioned, this feature is useful for migrating schema(which means no data) from other databases into Greenplum. 

I think the root cause of your wrong result is that the `create unique index on foopart_1_prt_1 (j);` makes the distribution key of the partition table and its child table to be different. We have two ideas to fix it.
1. Forbid it on non-root table, just like what alter table set distribution clause does.
```
create table foopart (i int4, j int4) distributed by (i) partition by
        range (i) (start (1) end (3) every (1));
alter table foopart_1_prt_1 set distributed by (j);
ERROR:  can't set the distribution policy of "foopart_1_prt_1"
HINT:  Distribution policy can be set for an entire partitioned table, not for one of its leaf parts or an interior branch.
```
2. Calculate the right flow and treat the partition table as randomly distributed. But this has bad performance impact. I prefer the first idea.

**Comment by gfphoenix78** [comment]
3 people in the above discussion are surprised by the behavior. Distribution keys are considered important in most cases. We shouldn't guess it for the users.

**Comment by zhangh43** [comment]
> 3 people in the above discussion are surprised by the behavior. Distribution keys are considered important in most cases. We shouldn't guess it for the users.

I prefer to at least use GUC to keep this feature(some customer may have already used this feature to do migration). If we fix the partition table issue, there is no wrong result for this existing feature.
Also you should keep all the test cases, not delete them directly. Also the error message should tell user how to create unique index if failed, e.g. `please update distributed key of the table before create unique index`

**Comment by zhangh43** [comment]
@gfphoenix78 

Another idea is to add the hint to tell user how to fix the failure when creating unique index on non distribution key. An similar example is alter table column type case as below. Our hint message could be `HINT: Alter table distribution key first to make it consistent with unique index columns`
```
demo=# create index j_idx on foo using btree (j);
demo=# alter table foo alter j type numeric;
ERROR:  cannot alter indexed column
HINT:  DROP the index first, and recreate it after the ALTER
```

File: src/test/regress/expected/create_table_like.out [hunk 1]
```diff
@@ -82,12 +82,11 @@ DETAIL:  Distribution key column "yy" is not included in the constraint.
 /* Ok to create multiple unique indexes */
 /* GPDB: This query will fail because unique index must contain all distribution key */
 CREATE TABLE inhg (x text UNIQUE, LIKE inhz INCLUDING INDEXES);
-ERROR:  UNIQUE constraint must contain all columns in the table's distribution key
-DETAIL:  Distribution key column "x" is not included in the constraint.
+ERROR:  UNIQUE or PRIMARY KEY definitions are incompatible with each other
+HINT:  When there are multiple PRIMARY KEY / UNIQUE constraints, they must have at least one column in common.
```
Comment (zhangh43) [inline src/test/regress/expected/create_table_like.out:86]: Here the error message changed. I think the original error `ERROR:  UNIQUE constraint must contain all columns in the table's distribution key` is correct, since `unique` make sense only when we could ensure they are distributed into the same segment.

File: src/test/regress/input/partindex_test.source [hunk 2]
```diff
@@ -369,10 +369,7 @@ create table part_table4
 -- create an index on all parts
 create index id_index on part_table4(id);
 
--- create 2 indexes on part_table4_1_prt_girls_2_prt_1
-create unique index id_index_unique on part_table4_1_prt_girls_2_prt_1(id);
```
Comment (zhangh43) [inline src/test/regress/input/partindex_test.source:373]: shall we update the distribution key instead of remove the test case?

File: src/test/regress/sql/gp_create_table.sql [hunk 3]
```diff
@@ -31,25 +31,6 @@ drop table distpol;
 create table distpol as select random(), 2 as foo distributed by (foo);
 select distkey, distclass from gp_distribution_policy where localoid = 'distpol'::regclass;
 drop table distpol;
--- now test that MPP-101 /actually/ works
-create table distpol (i int, j int, k int) distributed by (i);
-alter table distpol add primary key (j);
-select distkey, distclass from gp_distribution_policy where localoid = 'distpol'::regclass;
--- make sure we can't overwrite it
-create unique index distpol_uidx on distpol(k);
```
Comment (zhangh43) [inline src/test/regress/sql/gp_create_table.sql:39]: again, instead of removing the test case, better to use `alter table distribution key` to keep these test cases

File: src/backend/parser/parse_utilcmd.c [hunk 4]
```diff
@@ -2095,6 +2095,65 @@ transformDistributedBy(ParseState *pstate,
 		distrkeys = likeDistributedBy->keyCols;
 	}
 
+	/**
+	 * check for unique index.
+	 * If distrkeys is not determined by the above process,
+	 * we consider the most common columns in all unique indexes
+	 * as the distribution keys.
+	 */
+	foreach(lc, cxt->inh_indexes)
```
Comment (zhangh43) [inline src/backend/parser/parse_utilcmd.c:2104]: Can you explain more about this block of code? This seems unrelated to disable changing the distribution keys implicitly when creating unique index. Maybe separate it to another PR?

**Comment by zhangh43** [comment]
@gfphoenix78 we have reached the agreement to disable creating unique index when distribution key mismatch offline. Please fix the comments and let's make it merged.

File: src/test/regress/input/partindex_test.source (hunk 2 above)
Comment (gfphoenix78) [reply src/test/regress/input/partindex_test.source:373]: No, the distribution key can't be changed implicitly by creating a unique index. It has been tested somewhere. So, I remove it.

File: src/test/regress/expected/create_table_like.out (hunk 1 above)
Comment (gfphoenix78) [reply src/test/regress/expected/create_table_like.out:86]: The command involves two unique index constraints, one is `text` and the other is `yy` inherited from inhz. More than one UNIQUE/PRIMARY KEY index constraints should take their shared columns into account. If they share no columns, the distribution key is unavailable.
We still keep the feature to choose the distribution keys hinted by the UNIQUE/PRIMARY KEY index.

File: src/test/regress/sql/gp_create_table.sql (hunk 3 above)
Comment (gfphoenix78) [reply src/test/regress/sql/gp_create_table.sql:39]: I have added new tests in gp_index.sql. The following tests are unexpected.

File: src/backend/parser/parse_utilcmd.c (hunk 4 above)
Comment (zhangh43) [reply src/backend/parser/parse_utilcmd.c:2105]: Got it. Better to have a comment why most common columns is suitable for this case: we need to ensure tuples with the same value on columns of unique index to be placed in the same segment. And taking any column of a unique index as part of distributed key satisfy the above requirement.

Also please fix your `FIXME` in the commits.

File: src/backend/parser/parse_utilcmd.c [hunk 5]
```diff
@@ -2095,6 +2095,66 @@ transformDistributedBy(ParseState *pstate,
...
+				if (iparam && iparam->name)
+				{
+					IndexElem *distrkey = makeNode(IndexElem);
+					distrkey->name = iparam->name;
+					distrkey->opclass = NULL;
+					new_distrkeys = lappend(new_distrkeys, distrkey);
+				}
+			}
+		}
+		distrkeys = new_distrkeys;
```
Comment (zhangh43) [inline src/backend/parser/parse_utilcmd.c:2156]: If we got some distribution keys ahead(e.g. c1 and c2 by distributed by statement), this logic will remove the existing distribution keys if it's not in unique index.

File: src/backend/parser/parse_utilcmd.c [hunk 6]
```diff
@@ -2095,6 +2095,66 @@ transformDistributedBy(ParseState *pstate,
...
+	}
```
Comment (zhangh43) [inline src/backend/parser/parse_utilcmd.c:2157]: need empty line after }

File: src/backend/parser/parse_utilcmd.c (hunk 5 above)
Comment (gfphoenix78) [reply src/backend/parser/parse_utilcmd.c:2156]: Only the common columns are kept. This function is called only if the CREATE TABLE command doesn't specify the distribution keys explicitly.

File: src/backend/parser/parse_utilcmd.c (hunk 6 above)
Comment (gfphoenix78) [reply src/backend/parser/parse_utilcmd.c:2163]: Added.

File: src/backend/parser/parse_utilcmd.c (hunk 4 above)
Comment (gfphoenix78) [reply src/backend/parser/parse_utilcmd.c:2111]: Updated the comments. I can't provide an example when the condition `!iparam || !iparam->name` is true.

File: src/backend/parser/parse_utilcmd.c [hunk 7]
```diff
@@ -2095,6 +2095,73 @@ transformDistributedBy(ParseState *pstate,
...
+			continue;
+
+		if (distrkeys)
+		{
+			foreach(cell, index_stmt->indexParams)
+			{
+				IndexElem *iparam = lfirst(cell);
+				ListCell *dkcell;
+
+				if (!iparam || !iparam->name)
```
Comment (gfphoenix78) [inline src/backend/parser/parse_utilcmd.c:2128]: @xiong-gang @zhangh43 Could you give me some advice on how to handle the nullity here?

Comment (zhangh43) [reply src/backend/parser/parse_utilcmd.c:2128]: You could refer to `generateClonedIndexStmt`. The iparam->name could be NULL and corresponding iparam->expr is set. Better to add test cases for expression index as well and check whether we should consider iparam->expr at here.
```
        iparam = makeNode(IndexElem);

		if (AttributeNumberIsValid(attnum))
		{
			/* Simple index column */
			char	   *attname;

			attname = get_relid_attribute_name(indrelid, attnum);
			keycoltype = get_atttype(indrelid, attnum);

			iparam->name = attname;
			iparam->expr = NULL;
		}
		else
		{
			/* Expressional index */
			Node	   *indexkey;
			bool		found_whole_row;

			if (indexpr_item == NULL)
				elog(ERROR, "too few entries in indexprs list");
			indexkey = (Node *) lfirst(indexpr_item);
			indexpr_item = lnext(indexpr_item);

			/* Adjust Vars to match new table's column numbering */
			indexkey = map_variable_attnos(indexkey,
										   1, 0,
										   attmap, attmap_length,
										   &found_whole_row);

			/* As in transformTableLikeClause, reject whole-row variables */
			if (found_whole_row)
				ereport(ERROR,
						(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
						 errmsg("cannot convert whole-row table reference"),
						 errdetail("Index \"%s\" contains a whole-row table reference.",
								   RelationGetRelationName(source_idx))));

			iparam->name = NULL;
			iparam->expr = indexkey;

			keycoltype = exprType(indexkey);
		}
```

As for the NULL check here, I think it's OK, since we just need the most common columns here.

Comment (gfphoenix78) [reply src/backend/parser/parse_utilcmd.c:2128]: My understanding is that the index element could be either a column name or an expression. One and only one of them is not null. If the index element is not a column name, it should be skipped to compute the most common columns. For example, 
`create table t(i int, j int, k int) distributed by (i,j);`
`create unique index on t(i, func1(j));` The first index element is a name, the second index element is an expression. The set of distribution keys is not a sub-set of the column names in the index, so it violates the compatibility and finally it fails.
But `create unique index on t(i, j);` will success.

Comment (zhangh43) [reply src/backend/parser/parse_utilcmd.c:2128]: Yes, we should skip the expression index column. Just want to point out the case when `iparam -> name` is NULL.

**Review by zhangh43 — APPROVED** [review]
